### PR TITLE
add theming to Gutenberg components'

### PR DESCRIPTION
**Pull request by psealock**
We can create WooCommerce themed components from Gutenberg components by using [postcss-themes](https://github.com/WordPress/gutenberg/tree/master/packages/postcss-themes) at build time.

`postcss-themes` function reads SCSS with the `theme` keyword to create several variations. Starting with something like this:

```css
.my-class {
    background-color: theme( primary );
}
```

Into something this

```css
.my-class {
    background-color: #ccc;
}
body.theme-one .my-class {
    background-color: #ddd;
}
body.theme-two .my-class {
    background-color: #eee;
}
```

## A question for you, dear reviewer
The theme class is appended to the class list on the `body`, which is really convenient. However, this increases specificity which must be overcome when overriding Gutenberg styles. An example from this PR is `.woocommerce_page_wc-admin .woocommerce-dropdown-button`.

**Is having to document and deal with this an acceptable tradeoff?**

## Screenshots

screen shot 2018-07-30 at 8 59 44 pm

### RequireGutenments

1. Gutenberg Sass variables not exposed. This PR simply copies the necessary variables to build the button. Gutenberg developers are open to making the variables available publicly, p1532941539000239-slack-gutenberg

2. `postcss-themes` requires a default set of theme parameters, which I included in this PR. It means the default styles are loaded twice, so that will need to be fixed in Gutenberg.

### CODE REVIEW

File: client/components/dropdown-button/style.scss [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /** @format */
 
-.woocommerce-dropdown-button {
+.woocommerce_page_wc-admin .woocommerce-dropdown-button {
```
Comment (justinshreve) [inline client/components/dropdown-button/style.scss]: If the button is embedded in the activity panel on a page like the orders or coupons page, I'm not sure this rule will work. Does the `woocommerce-page` class added [here](https://github.com/woocommerce/wc-admin/blob/c361948bc3b4c01bfd8e5f38cd3f98a4e77527bd/lib/admin.php#L147) work?

File: client/stylesheets/_variables.scss [hunk 2]
```diff
@@ -9,3 +9,21 @@ $gap-smallest: 4px;
...
+
+// Gutenberg Button variables. So very temporary
+$default-font-size: 16px;
+$blue-medium-200: #bfe7f3;
+$light-gray-500: #e2e4e7;
+$dark-gray-300: #6c7781;
+$dark-gray-900: #191e23;
+$alert-red: #d94f4f;
+
+@mixin button-style__focus-active() {
```
Comment (justinshreve) [inline client/stylesheets/_variables.scss]: Is this mixin temporary as well? If not, it should go in the mixins file instead.

File: client/stylesheets/_variables.scss [hunk 3]
```diff
@@ -9,3 +9,21 @@ $gap-smallest: 4px;
 
 // @todo remove this spacing variable
 $spacing: 16px;
+
+// Gutenberg Button variables. So very temporary
+$default-font-size: 16px;
```
Comment (justinshreve) [inline client/stylesheets/_variables.scss]: Although these are temporary -- all but the `$blue-medium` variables are either defined in `_colors.scss` or via the font-size mixin.

File: webpack.config.js [hunk 4]
```diff
@@ -53,12 +53,43 @@ const webpackConfig = {
 					fallback: 'style-loader',
 					use: [
 						'css-loader',
+						{
+							loader: 'postcss-loader',
+							options: {
+								plugins: [
+									require( './node_modules/@wordpress/postcss-themes' )( {
```
Comment (justinshreve) [inline webpack.config.js]: I merged a change that adds `autoprefixer` and `postcss-loader`. These changes can actually go in [`postcss.config.js`](https://github.com/woocommerce/wc-admin/blob/04e6ff6cdf8cb99879d98c7f9d82d4f8afb3198f/postcss.config.js) now instead.

**Review by emdashcodes — COMMENTED** [review]
The font on the upgrade and reset buttons are a little big when testing, but this seems to be working.

> Is having to document and deal with this an acceptable tradeoff?

I don't think it's a huge deal if we only need to do it when overwriting some Gutenberg styles. I wonder if there is anything we could do to make it easier though? Maybe a mixin -- or could we auto append `woocommerce-page` around things using the build process? (p4TIVU-94S-p2 "Simultaneously we will be developing a webpack loader for CSS which post-processes the generated styles and prefixes a class name of choice to all CSS rules.")

**Comment by psealock** [comment]
Thanks @justinshreve. Here are some changes:

* configs move to `postcss.config.js`
* reference variables in `_colors.scss`
* move the mixin to the mixin file
* changed the class to `woocommerce-page`
* ensured the new styles are available on embedded app

**Review by emdashcodes — APPROVED** [review]
Thanks for the changes. Looking good to me and testing well.

**Comment by ryelle** [comment]
I'm not familiar with postcss-themes, so I have a bunch of questions 🙂 

Right now you're only running this on gutenberg's button CSS, right? You haven't updated any of our code to use these theme vars. Do you think we'll need to?

All the colors in `woocommerce-page` are the same right now, was that just to get this PR started? I wouldn't expect `primary` and `secondary` to be the same color, for example.

We'll probably want to drop the gutenberg SCSS imports into their own file, since we'll need to import every component's sass file, and maybe include that in `_wpadmin-reset.scss` (which has already started to become more than just a reset…) - then it'll be in both embedded and the full app.

> Gutenberg Sass variables not exposed

~By this you mean that we can't use the sass variables in our CSS? Does that matter?~ Aha, I just tried `import 'gutenberg-components/style.scss';` and I see why this is a problem - and why you've added the variables & `button-style__focus-active()`.

**Comment by psealock** [comment]
Thanks for taking a look and asking thoughtful questions @ryelle.

>Right now you're only running this on gutenberg's button CSS, right? You haven't updated any of our code to use these theme vars. Do you think we'll need to?

Correct, I'm only importing styles for the button. I figure we can do it as needed. 

>You haven't updated any of our code to use these theme vars. Do you think we'll need to?

Our code has not been changed, nor do I think it will need to. I'm simply introducing additional CSS rules to existing Gutenberg CSS with higher specificity, eg `body.woocommerce-page .components-button`. I suppose if we want to introduce themeing on our project, that could change.

>All the colors in woocommerce-page are the same right now, was that just to get this PR started? I wouldn't expect primary and secondary to be the same color, for example.

yes, good spot. I was just making them the same for now, but need to revisit that. For instance, `secondary` is likely the same colour as Gutenberg's. I'll look into that today.

>We'll probably want to drop the gutenberg SCSS imports into their own file, since we'll need to import every component's sass file

Thats a good idea, I'll look into `_wpadmin-reset.scss` as a potential home.

**Comment by psealock** [comment]
I moved the imports to `_wpadmin-reset.scss` as suggested, which allows us to import once instead of in `index.js` and `embedded.js`.

I feel ok leaving all the config colors the same for now. For reference, here are the defaults:

screen shot 2018-08-01 at 11 07 33 am

They are all similar shades of blue and considering much of the `theme` function uses are shaded anyways, I don't think it matters much at this point.

```css
color( theme( button ) shade( 20% ) )
```

**Comment by psealock** [comment]
I'll merge this tomorrow unless there are more questions or comments

**Pull request merged**